File: src/commands/verify-log.ts
import { verifyAuditLogChain } from "../audit-log/chain-verifier.js";
import { DEFAULT_AUDIT_LOG_PATH } from "../audit-log/hash-chain-writer.js";
import type { ChainVerificationResult } from "../types.js";

export interface VerifyLogCommandArgs {
  path?: string;
  json?: boolean;
}

export interface VerifyLogResult extends ChainVerificationResult {
  logPath: string;
}

/**
 * The programmatic core of `verify-log`: re-derives the chain and returns
 * the structured result, with no console/stdout output of its own. Shared
 * by the CLI's `--json` mode and the MCP `verify_log` tool.
 */
export async function executeVerifyLog(path?: string): Promise<VerifyLogResult> {
  const logPath = path ?? DEFAULT_AUDIT_LOG_PATH;
  const result = await verifyAuditLogChain(logPath);
  return { logPath, ...result };
}

export async function runVerifyLogCommand(args: VerifyLogCommandArgs): Promise<void> {
  const logPath = args.path ?? DEFAULT_AUDIT_LOG_PATH;
  if (!args.json) {
    console.log(`Verifying ${logPath}...`);
  }

  const result = await executeVerifyLog(args.path);

  if (!result.valid) {
    process.exitCode = 1;
  }

  if (args.json) {
    process.stdout.write(`${JSON.stringify(result, null, 2)}\n`);
    return;
  }

  if (result.totalEntries === 0) {
    console.log("No entries yet -- nothing to verify.");
    return;
  }

  if (result.valid) {
    console.log(`✓ Chain intact: ${result.totalEntries} entries, no gaps, no tampering detected.`);
  } else {
    console.error(
      `✗ Chain broken at entry ${result.brokenAtIndex} (${result.brokenAtEntryId ?? "unknown"}): ${result.reason}`,
    );
  }
}

File: python/src/auditreach/commands/verify_log.py
"""Ported from src/commands/verify-log.ts."""
from __future__ import annotations

import dataclasses
import json
import sys
from typing import Optional

from ..audit_log.chain_verifier import verify_audit_log_chain
from ..audit_log.hash_chain_writer import DEFAULT_AUDIT_LOG_PATH


def run_verify_log_command(path: Optional[str] = None, json_output: bool = False) -> int:
    """Returns the process exit code (0 valid / 1 tampered)."""
    log_path = path or DEFAULT_AUDIT_LOG_PATH
    if not json_output:
        print(f"Verifying {log_path}...")

    result = verify_audit_log_chain(log_path)
    exit_code = 0 if result.valid else 1

    if json_output:
        payload = {"logPath": log_path, **dataclasses.asdict(result)}
        sys.stdout.write(json.dumps(payload, indent=2) + "\n")
        return exit_code

    if result.total_entries == 0:
        print("No entries yet -- nothing to verify.")
        return exit_code

    if result.valid:
        print(f"✓ Chain intact: {result.total_entries} entries, no gaps, no tampering detected.")
    else:
        print(
            f"✗ Chain broken at entry {result.broken_at_index} "
            f"({result.broken_at_entry_id or 'unknown'}): {result.reason}",
            file=sys.stderr,
        )

    return exit_code
